apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -498,39 +511,87 @@ public Flow(long maxByteCount, long maxElementCount, Action<Task> registerTaskFn
+                                    return;
+                                }
+                                // Mark this ordering-key inflight, and allow user code for this first message to be executed.
+                                // Set the queue to null to save allocation.
+                                // It'll be created on-demand if more than one message with the same key are queued.
+                                _keyedTaskQs.Add(orderingKey, null);
+                            }
                             break;
                         }
                     }
                     // Flow not OK, wait until a previous fn completes.
                     // CancellationToken not required, as the fn() will always drain on cancellation.
                     await _taskHelper.ConfigureAwait(_event.WaitAsync(CancellationToken.None));
                 }
+                // Execute the user code for this message.
+                ExecuteFunction(orderingKey, byteCount, fn);
+            }
+
+            private void ExecuteFunction(string orderingKey, long byteCount, Func<Task> fn)
+            {
                 // Execute fn, and schedule the following code to execute once it has completed.
                 // Register the function, so we can be sure it's completed during shutdown.
                 Task task = _taskHelper.Run(async () => await _taskHelper.ConfigureAwaitWithFinally(fn, () =>
                 {
                     bool setEvent;
+                    FnInfo nextFn = default;
                     lock (_lock)
                     {
+                        if (orderingKey.Length > 0)
+                        {
+                            var taskQ = _keyedTaskQs[orderingKey];
+                            if (taskQ != null && taskQ.Count > 0)
+                            {
+                                // More fn(s) to execute with this same ordering-key; execute the next fn.
+                                nextFn = taskQ.Dequeue();
+                            }
+                            else
+                            {
+                                // All fns executed for this ordering-key; remove the key.
+                                _keyedTaskQs.Remove(orderingKey);
+                            }
+                        }
                         bool preIsFlowOk = IsFlowOk();
                         _byteCount -= byteCount;
                         _elementCount -= 1;
                         // If moving from flow-bad to flow-OK, then trigger the next execution.
                         setEvent = !preIsFlowOk && IsFlowOk();
                     }
+                    if (nextFn.Fn != null)
+                    {
+                        // Execute user code for the next message of this ordering-key.
+                        ExecuteFunction(orderingKey, nextFn.ByteCount, nextFn.Fn);
```

Comment:
Possibly add a comment to explain that this isn't really recursing due to Task.Run? Otherwise it looks dangerous from a stack perspective.

Reply:
Done :)